### PR TITLE
(extensions) probes: create probe group, events: start work

### PR DESCRIPTION
**Pull request by rafaeldtinoco**
This is the initial work of the Tracee Extensions, I'm preparing the baseline for https://github.com/aquasecurity/tracee/issues/3170 (the multiple eBPF objects), which is a needed step for https://github.com/aquasecurity/tracee/issues/2991 (API server) and the extensions:
    
    Each extension will consist of:
    
    1. Map Group    - eBPF maps used by this extension
    2. Probe Group  - eBPF probes used by this extension
    3. Functions    - tracee hooks callbacks
    4. Event Group  - events declared by this extension
    5. Signature(s) - signatures declared by this extension
    
    Logic and packages will change. This is Work in Progress.

commit 736137d5 (HEAD -> extensions-probe-group, rafaeldtinoco/extensions-probe-group, myowntracee)
Author: Rafael David Tinoco <rafaeldtinoco@gmail.com>
Date:   Mon Jun 12 04:53:14 2023

    events: tailcall dependencies should be instanced and thread safe
    
    related: #3170

commit ef781442
Author: Rafael David Tinoco <rafaeldtinoco@gmail.com>
Date:   Mon Jun 12 02:02:42 2023

    events: probes are event dependencies already
    
    ... make it so, then.
    
    related: #3170

commit 7d588e99
Author: Rafael David Tinoco <rafaeldtinoco@gmail.com>
Date:   Mon Jun 5 19:16:49 2023

    probes: create probe group
    
    - refactors probes
    - creates probe group
    - makes sure probe group is thread safe
    
    related: #3170

### CODE REVIEW

File: pkg/ebpf/tracee.go
Comment (NDStrahilevitz) [inline pkg/ebpf/tracee.go:1144]: nit: Let's start "For syscall tracing..." on a new line as it is a different subject.

File: pkg/events/events.go
Comment (NDStrahilevitz) [inline pkg/events/events.go:84]: I don't see any usage of this function except in `NewTailCallFull` below. Any reason not to merge them into `NewTailCall`?
Also, maybe we should introduce a `tailcall` package for type separation and to reduce this file size. WDYT?

File: pkg/ebpf/probes/probe_group.go
Comment (rafaeldtinoco) [reply pkg/ebpf/probes/probe_group.go:18]: Yep... I wanted to highlight they're a big lock guaranteeing that there won't be 2 attachments, from different probes, within the same probe group, happening simultaneously (more than just protecting probes iterations).

File: pkg/events/events.go
Comment (NDStrahilevitz) [reply pkg/events/events.go:84]: I'm not sure there's any meaning to initializing a tail call with no indices... If I wanted to construct such a tail call (for some reason) I would pass an empty array. Even then, for the tail call to do anything, it would need to be given some index eventually.
Your call but I think you won't need the "non-full" version.

File: pkg/events/events.go
Comment (rafaeldtinoco) [reply pkg/events/events.go:99]: Your arguments are fair. I have removed the full version.

Comment (rafaeldtinoco) [reply pkg/events/events.go:99]: I thought about having mapIndexes at the end as well, so, since you also suggests, I'm making that change as well.

**Comment by rafaeldtinoco** [comment]
> +1ing and it's your call if you want to make the following changes before merging.

You had good and valid points. I did the changes and pushed, I'll consider you're still +1 as there are no other changes. Will merge if tests pass after the current bug fix release.

Thanks for reviewing this!

**Review by josedonizetti — APPROVED** [review]
Some typos only @rafaeldtinoco, LGMT otherwise.

**Comment by yanivagman** [comment]
Forgot to write that I want to review this before merging...
Will try to go over it today

File: pkg/events/events.go
Comment (yanivagman) [inline pkg/events/events.go:49]: Bad alignment

**Comment by rafaeldtinoco** [comment]
> Forgot to write that I want to review this before merging...

Thanks for reviewing, no matter the timing. I'll provide a chore PR with adjustments.

File: pkg/events/events.go
Comment (rafaeldtinoco) [reply pkg/events/events.go:22]: This way we don't have to do things as:

- keep data related to event definitions externally to the interface
- synchronize data got from the definitions externally (and not centralized)

example of that would be:

- pick a list of dependencies for an event
- manipulate that list
- save the list somewhere (like in the tracee singleton)

and now interface users have:

- event definitions static list
- stateful information about dependencies in a singleton
- possible other interface users saving state elsewhere

if we instantiate the event definitions we can:

- centralize field protection
- guarantee transactions
- turn everything concurrent from a single point of view

and even if that all wasn't true, the question would be "why not ?", as keeping instances for event dependencies wouldn't cause any harm EVEN if not needed (which isn't the case).

File: pkg/ebpf/tracee.go
Comment (rafaeldtinoco) [reply pkg/ebpf/tracee.go:1194]: Its a single step to make it all dynamic and thread-safe. I started with tailcalls, everything else will be instanciable and protected. At the end I'll run locking contention tests to see if we should change locking patterns (making things atomic) or even remove locks if they're not being used.

File: pkg/ebpf/probes/probe_group.go
Comment (rafaeldtinoco) [reply pkg/ebpf/probes/probe_group.go:48]: its another probe type (probes/cgroup.go), these are for the "TraceProbe" only.

File: pkg/events/events.go
Comment (rafaeldtinoco) [reply pkg/events/events.go:22]: Ok I think we got into an agreement in our offline conversation (meeting) for this item.

**Comment by rafaeldtinoco** [comment]
Including observations in the next PR.

**Pull request merged**
